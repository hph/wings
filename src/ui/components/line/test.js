import React from 'react';
import renderer from 'react-test-renderer';

import { Line, mapTokens, tokenizerOptions, collectTokens } from './index';

jest.mock('ui/components/token', () => 'Token');

const sampleProgram = `
import React from 'react';

export default class SomeComponent extends React.Component {
  state = {
    isMounted: false,
  };

  componentDidMount () {
    console.log('Mounted!');
    this.setSTate({ isMounted: true });
  }

  render () {
    return (
      <div>Component {!this.state.isMounted && 'not'} mounted.</div>
    );
  }
}
`.trim();

describe('tokenizer', () => {
  describe('tokenizerOptions', () => {
    it('should have a static value', () => {
      expect(tokenizerOptions).toEqual({
        plugins: {
          jsx: true,
        },
        loc: true,
      });
    });
  });

  describe('collectTokens', () => {
    it('should return an empty array in the event of an error', () => {
      expect(collectTokens('const foo = \'')).toEqual([]);
    });

    it('should output "let" as the appropriate token', () => {
      expect(collectTokens('let foo = \'bar\';')).toMatchSnapshot();
    });

    it('should output "from" as the appropriate token', () => {
      expect(collectTokens('import React from \'react\';')).toMatchSnapshot();
    });

    it('should output the appropriate token for template tags', () => {
      expect(collectTokens('`foo`')).toMatchSnapshot();
    });

    it('should output the appropriate tokens for JSX tags', () => {
      expect(collectTokens('const Foo = () => <div>foo</div>;')).toMatchSnapshot();
    });

    it('should output the appropriate tokens for RegEx patterns', () => {
      expect(collectTokens('const pattern = /pat/;')).toMatchSnapshot();
    });

    it('should output the appropriate tokens for comments', () => {
      expect(collectTokens('const foo = \'bar\'; // yep!')).toMatchSnapshot();
      expect(collectTokens('/* This is a comment */')).toMatchSnapshot();
    });

    it('should output the appropriate tokens for a more complex program', () => {
      expect(collectTokens(sampleProgram)).toMatchSnapshot();
    });
  });
});

describe('Line', () => {
  it('should render children if no tokens are provided', () => {
    expect(
      renderer.create(<Line tokens={[]}>Hello, world!</Line>).toJSON(),
    ).toMatchSnapshot();
  });

  it('should render tokens if provided', () => {
    const tokens = [
      {
        type: 'declaration',
        value: 'const',
      },
      {
        value: ' ',
      },
      {
        type: 'name',
        value: 'foo',
      },
      {
        type: 'operator',
        value: '=',
      },
      {
        type: 'value',
        value: '\'bar\'',
      },
      {
        type: 'comment',
        value: ';',
      },
    ];
    expect(
      renderer.create(
        <Line tokens={tokens}>
          {'const foo = \'bar\';'}
        </Line>,
      ).toJSON(),
    ).toMatchSnapshot();
  });
});

describe('mapTokens', () => {
  it('should collect tokens from children and provide them as a prop', () => {
    const props = mapTokens({ children: 'const foo = \'bar\';' });
    expect(props).toEqual({
      children: 'const foo = \'bar\';',
      tokens: [{
        type: 'declaration',
        value: 'const',
      }, {
        value: ' ',
      }, {
        type: 'name',
        value: 'foo',
      }, {
        value: ' ',
      }, {
        type: 'operator',
        value: '=',
      }, {
        value: ' ',
      }, {
        type: 'value',
        value: '\'bar\'',
      }, {
        type: 'comment',
        value: ';',
      }],
    });
  });
});
