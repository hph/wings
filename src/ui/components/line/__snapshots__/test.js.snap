// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line should render children if no tokens are provided 1`] = `
<div>
  Hello, world!
</div>
`;

exports[`Line should render tokens if provided 1`] = `
<div>
  <Token
    type="declaration"
  >
    const
  </Token>
  <Token
    type={undefined}
  >
     
  </Token>
  <Token
    type="name"
  >
    foo
  </Token>
  <Token
    type="operator"
  >
    =
  </Token>
  <Token
    type="value"
  >
    'bar'
  </Token>
  <Token
    type="comment"
  >
    ;
  </Token>
</div>
`;

exports[`tokenizer collectTokens should output "from" as the appropriate token 1`] = `
Array [
  Object {
    "type": "module",
    "value": "import",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "React",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "module",
    "value": "from",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "'react'",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
]
`;

exports[`tokenizer collectTokens should output "let" as the appropriate token 1`] = `
Array [
  Object {
    "type": "declaration",
    "value": "let",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "foo",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "operator",
    "value": "=",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "'bar'",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
]
`;

exports[`tokenizer collectTokens should output the appropriate token for template tags 1`] = `
Array [
  Object {
    "type": "value",
    "value": "\`",
  },
  Object {
    "type": "value",
    "value": "foo",
  },
  Object {
    "type": "value",
    "value": "\`",
  },
]
`;

exports[`tokenizer collectTokens should output the appropriate tokens for JSX tags 1`] = `
Array [
  Object {
    "type": "declaration",
    "value": "const",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "Foo",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "operator",
    "value": "=",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "(",
  },
  Object {
    "type": "other",
    "value": ")",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "operator",
    "value": "=>",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "jsx",
    "value": "<",
  },
  Object {
    "type": "jsx",
    "value": "div",
  },
  Object {
    "type": "jsx",
    "value": ">",
  },
  Object {
    "type": "name",
    "value": "foo",
  },
  Object {
    "type": "jsx",
    "value": "<",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
  Object {
    "type": "jsx",
    "value": "div",
  },
  Object {
    "type": "jsx",
    "value": ">",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
]
`;

exports[`tokenizer collectTokens should output the appropriate tokens for RegEx patterns 1`] = `
Array [
  Object {
    "type": "declaration",
    "value": "const",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "pattern",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "operator",
    "value": "=",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "/pat/",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
]
`;

exports[`tokenizer collectTokens should output the appropriate tokens for a more complex program 1`] = `
Array [
  Object {
    "type": "module",
    "value": "import",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "React",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "module",
    "value": "from",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "'react'",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
  Object {
    "value": "  ",
  },
  Object {
    "type": "module",
    "value": "export",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "keyword",
    "value": "default",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "keyword",
    "value": "class",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "SomeComponent",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "keyword",
    "value": "extends",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "React",
  },
  Object {
    "type": "other",
    "value": ".",
  },
  Object {
    "type": "name",
    "value": "Component",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "{",
  },
  Object {
    "value": "   ",
  },
  Object {
    "type": "name",
    "value": "state",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "operator",
    "value": "=",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "{",
  },
  Object {
    "value": "     ",
  },
  Object {
    "type": "name",
    "value": "isMounted",
  },
  Object {
    "type": "operator",
    "value": ":",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "false",
  },
  Object {
    "type": "other",
    "value": ",",
  },
  Object {
    "value": "   ",
  },
  Object {
    "type": "other",
    "value": "}",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
  Object {
    "value": "    ",
  },
  Object {
    "type": "name",
    "value": "componentDidMount",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "(",
  },
  Object {
    "type": "other",
    "value": ")",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "{",
  },
  Object {
    "value": "     ",
  },
  Object {
    "type": "name",
    "value": "console",
  },
  Object {
    "type": "other",
    "value": ".",
  },
  Object {
    "type": "name",
    "value": "log",
  },
  Object {
    "type": "other",
    "value": "(",
  },
  Object {
    "type": "value",
    "value": "'Mounted!'",
  },
  Object {
    "type": "other",
    "value": ")",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
  Object {
    "value": "     ",
  },
  Object {
    "type": "keyword",
    "value": "this",
  },
  Object {
    "type": "other",
    "value": ".",
  },
  Object {
    "type": "name",
    "value": "setSTate",
  },
  Object {
    "type": "other",
    "value": "(",
  },
  Object {
    "type": "other",
    "value": "{",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "isMounted",
  },
  Object {
    "type": "operator",
    "value": ":",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "true",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "}",
  },
  Object {
    "type": "other",
    "value": ")",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
  Object {
    "value": "   ",
  },
  Object {
    "type": "other",
    "value": "}",
  },
  Object {
    "value": "    ",
  },
  Object {
    "type": "name",
    "value": "render",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "(",
  },
  Object {
    "type": "other",
    "value": ")",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "{",
  },
  Object {
    "value": "     ",
  },
  Object {
    "type": "keyword",
    "value": "return",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "(",
  },
  Object {
    "value": "       ",
  },
  Object {
    "type": "jsx",
    "value": "<",
  },
  Object {
    "type": "jsx",
    "value": "div",
  },
  Object {
    "type": "jsx",
    "value": ">",
  },
  Object {
    "type": "name",
    "value": "Component ",
  },
  Object {
    "type": "other",
    "value": "{",
  },
  Object {
    "type": "other",
    "value": "!",
  },
  Object {
    "type": "keyword",
    "value": "this",
  },
  Object {
    "type": "other",
    "value": ".",
  },
  Object {
    "type": "name",
    "value": "state",
  },
  Object {
    "type": "other",
    "value": ".",
  },
  Object {
    "type": "name",
    "value": "isMounted",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "operator",
    "value": "&&",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "'not'",
  },
  Object {
    "type": "other",
    "value": "}",
  },
  Object {
    "type": "name",
    "value": " mounted.",
  },
  Object {
    "type": "jsx",
    "value": "<",
  },
  Object {
    "type": "operator",
    "value": "/",
  },
  Object {
    "type": "jsx",
    "value": "div",
  },
  Object {
    "type": "jsx",
    "value": ">",
  },
  Object {
    "value": "     ",
  },
  Object {
    "type": "other",
    "value": ")",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
  Object {
    "value": "   ",
  },
  Object {
    "type": "other",
    "value": "}",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "other",
    "value": "}",
  },
]
`;

exports[`tokenizer collectTokens should output the appropriate tokens for comments 1`] = `
Array [
  Object {
    "type": "declaration",
    "value": "const",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "name",
    "value": "foo",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "operator",
    "value": "=",
  },
  Object {
    "value": " ",
  },
  Object {
    "type": "value",
    "value": "'bar'",
  },
  Object {
    "type": "comment",
    "value": ";",
  },
  Object {
    "type": "comment",
    "value": " // yep!",
  },
]
`;

exports[`tokenizer collectTokens should output the appropriate tokens for comments 2`] = `
Array [
  Object {
    "type": "comment",
    "value": "/* This is a comment */",
  },
]
`;
